scrape/scrape.go

```diff
@@ -510,44 +518,38 @@ func (sp *scrapePool) Sync(tgs []*targetgroup.Group) {
 // scrape loops for new targets, and stops scrape loops for disappeared targets.
 // It returns after all stopped scrape loops terminated.
 func (sp *scrapePool) sync(targets []*Target) {
-	var (
-		uniqueLoops   = make(map[uint64]loop)
-		interval      = time.Duration(sp.config.ScrapeInterval)
-		timeout       = time.Duration(sp.config.ScrapeTimeout)
-		bodySizeLimit = int64(sp.config.BodySizeLimit)
-		sampleLimit   = int(sp.config.SampleLimit)
-		labelLimits   = &labelLimits{
-			labelLimit:            int(sp.config.LabelLimit),
-			labelNameLengthLimit:  int(sp.config.LabelNameLengthLimit),
-			labelValueLengthLimit: int(sp.config.LabelValueLengthLimit),
-		}
-		honorLabels     = sp.config.HonorLabels
-		honorTimestamps = sp.config.HonorTimestamps
-		mrc             = sp.config.MetricRelabelConfigs
-	)
+	uniqueLoops := make(map[uint64]loop)
 
 	sp.targetMtx.Lock()
 	for _, t := range targets {
 		hash := t.hash()
 
 		if _, ok := sp.activeTargets[hash]; !ok {
-			// The scrape interval and timeout labels are set to the config's values initially,
-			// so whether changed via relabeling or not, they'll exist and hold the correct values
-			// for every target.
-			var err error
-			interval, timeout, err = t.intervalAndTimeout(interval, timeout)
-
-			s := &targetScraper{Target: t, client: sp.client, timeout: timeout, bodySizeLimit: bodySizeLimit}
+			var (
+				err error
+				// The scrape interval, timeout and various limits are set to the config's values initially,
+				// so whether changed via relabeling or not, they'll exist and hold the correct values
+				// for every target.
+				targetSettings *labelSettings
+			)
+			targetSettings, err = t.labelSettings(scrapeConfigDefaults(sp.config))
```

Comment:
In here we parse labels for the second time to use in `setForcedError`, as was already done in `t.intervalAndTimeout`.

Before https://github.com/prometheus/prometheus/pull/8911 all validation was performed during the relabel phase.
I propose to keep validation in `PopulateLabels` altogether and propagate `labelSettings` via return argument. It may be embedded in `Target` struct later. And in here, while creating a scraper, we would just use `labelSettings` that were already validated.